packages/@react-native-windows/codegen/Document.md

```diff
@@ -0,0 +1,386 @@
+{
+    result.Reject("error!");
+    // accepting char const*, wchar_t const* and winrt::Microsoft::ReactNative::ReactError
+}
+```
+
+### REACT_GET_CONSTANTS
+
+`REACT_GET_CONSTANTS` implements the sync method `getConstants` in a TurboModule.
+`getConstants` returns an object literal without any argument.
+
+The syntax of `REACT_GET_CONSTANTS` is `REACT_GET_CONSTANTS(method)`.
+
+- `method`: the C++ method to call. Whatever the name is, it becomes `getConstants` in JavaScript.
+
+Although they are semantically the same in JavaScript but `REACT_GET_CONSTANTS(method)` could never be replaced by `REACT_SYNC_METHOD(method, getConstants)`.
+
+```c++
+REACT_GET_CONSTANTS(GetConstants)
+MyStruct GetConstants() noexcept {
+    return { ... };
+}
+
+REACT_GET_CONSTANTS(GetConstantsStatic)
+static MyStruct GetConstants() noexcept {
+    return { ... };
+}
+```
+
+**Important**: If multiple `REACT_GET_CONSTANTS`, `REACT_CONSTANT_PROVIDER` and `REACT_CONSTANT` appear in a module struct,
+calling `getConstants` in JavaScript gets the result of all these methods combined together.
+
+**Important**: If `getConstants` are in the generated spec class,
+then a single `REACT_GET_CONSTANTS` becomes the only way to implement this member.
+
+### REACT_CONSTANT_PROVIDER
+
+`REACT_CONSTANT_PROVIDER` is a weak-typed version of `REACT_GET_CONSTANTS`.
+The attached function must returns `void`
+and its only argument must be `winrt::Microsoft::ReactNative::ReactConstantProvider& provider`.
+
+The syntax of `REACT_CONSTANT_PROVIDER` is `REACT_CONSTANT_PROVIDER(method)`.
+
+- `method`: the C++ method to call. Whatever the name is, it becomes `getConstants` in JavaScript.
+
+```c++
+REACT_CONSTANT_PROVIDER(GetConstants)
+void GetConstants(winrt::Microsoft::ReactNative::ReactConstantProvider& provider) noexcept
+{
+    provider.Add(L"name", value);
+    ...
+}
+
+REACT_CONSTANT_PROVIDER(GetConstantsStatic)
+static void GetConstants(winrt::Microsoft::ReactNative::ReactConstantProvider& provider) noexcept
+{
+    provider.Add(L"name", value);
+    ...
+}
+```
+
+**Important**: If multiple `REACT_GET_CONSTANTS`, `REACT_CONSTANT_PROVIDER` and `REACT_CONSTANT` appear in a module struct,
+calling `getConstants` in JavaScript gets the result of all these methods combined together.
+
+### REACT_CONSTANT
+
+`REACT_CONSTANT`
+
+`REACT_CONSTANT` adds a field to `getConstants`.
+
+The syntax of `REACT_CONSTANT` is `REACT_CONSTANT(field)`.
+
+- `field`: the C++ field to read.
+
+```c++
+REACT_CONSTANT(Constant1)
+const std::string Constant1 {"Something"};
+
+REACT_CONSTANT(Constant2)
+static constexpr int Constant2 = 100;
+```
+
+**Important**: If multiple `REACT_GET_CONSTANTS`, `REACT_CONSTANT_PROVIDER` and `REACT_CONSTANT` appear in a module struct,
+calling `getConstants` in JavaScript gets the result of all these methods combined together.
+
+### REACT_EVENT
+
+`REACT_EVENT` implements an event.
+`REACT_EVENT` attaches to a `std::function` field.
```

Comment:
>std::function [](http://example.com/codeflow?start=29&length=13)

Technically we support any std::function - like template class. I.e. the type must accept a method signature as a template type parameter.

Reply:
I do mean it. I think just writing `std::function` in C++ does not compile. and it must have a template argument. People should be able to catch that.

Reply:
My point is that there is no requirement to use std::function. We can use Mso::Functor or any other type that has the same class template shape.